File: examples/29-valgrind-sanitizers-miri/go/main.go
// bugfarm — chapter 29: valgrind, sanitizers, and miri.
//
//	app <leak|uaf|uninit|overflow|race>
//
// Go is memory-safe by construction: there's no dangling pointer to chase
// after free (the GC never reclaims memory a live reference still points
// at), no read of uninitialised memory (every value starts zeroed), and no
// out-of-bounds write that silently corrupts a neighbor (a slice/array
// access past its bound panics immediately -- a runtime check, not
// something a background tool has to catch after the fact). So uaf,
// uninit, and overflow print a "prevented by the runtime" note and exit
// 64: that absence of a bug *is* the language's story here.
//
// What Go can still get wrong:
//
//	leak -- a goroutine that blocks forever. There's no valgrind
//	        equivalent watching the Go heap for this; the way you notice
//	        is runtime.NumGoroutine() not going back down.
//	race -- unsynchronized access to a shared variable from goroutines.
//	        Build with `go build -race` and the runtime's race detector
//	        prints "WARNING: DATA RACE" and the process exits 66.
package main

import (
	"fmt"
	"os"
	"runtime"
	"sync"
	"time"
)

const usage = "usage: app <leak|uaf|uninit|overflow|race>"
const prevented = "go: prevented by the runtime — see chapter"

func parseBug(arg string) (string, error) {
	switch arg {
	case "leak", "uaf", "uninit", "overflow", "race":
		return arg, nil
	default:
		return "", fmt.Errorf("unknown bug: %s", arg)
	}
}

// leak: spawns a goroutine that blocks forever on a channel receive nobody
// will ever send to. Nothing external is watching for this; we notice it
// the same way a real operator would, by the goroutine count never coming
// back down.
func leak() int {
	before := runtime.NumGoroutine()
	block := make(chan struct{})
	go func() {
		<-block // never sent to: this goroutine (and its stack) leaks forever
	}()
	time.Sleep(50 * time.Millisecond) // let the scheduler actually start it
	after := runtime.NumGoroutine()
	leaked := after - before
	fmt.Printf("bugfarm: leak: goroutines before=%d after=%d leaked=%d\n", before, after, leaked)
	if leaked < 1 {
		fmt.Fprintln(os.Stderr, "bugfarm: leak: expected goroutine leak was not observed")
		return 1
	}
	return 0
}

var counter int // race: shared, unsynchronized on purpose

func bump(wg *sync.WaitGroup) {
	defer wg.Done()
	for i := 0; i < 100_000; i++ {
		counter++
	}
}

// race: two goroutines increment a shared int with no lock or atomic. Under
// a plain build this just risks a wrong final count; built with -race, the
// runtime instruments every access and reports the conflicting pair.
func race() int {
	var wg sync.WaitGroup
	wg.Add(2)
	go bump(&wg)
	go bump(&wg)
	wg.Wait()
	fmt.Printf("bugfarm: race: counter=%d (expected 200000; a wrong value is the benign symptom)\n",
		counter)
	return 0
}

func main() {
	if len(os.Args) != 2 {
		fmt.Fprintln(os.Stderr, usage)
		os.Exit(2)
	}
	bug, err := parseBug(os.Args[1])
	if err != nil {
		fmt.Fprintf(os.Stderr, "app: %v\n%s\n", err, usage)
		os.Exit(2)
	}
	switch bug {
	case "leak":
		os.Exit(leak())
	case "race":
		os.Exit(race())
	default: // uaf, uninit, overflow: not expressible in safe Go
		fmt.Fprintln(os.Stderr, prevented)
		os.Exit(64)
	}
}

File: examples/29-valgrind-sanitizers-miri/rust/src/main.rs
//! bugfarm — chapter 29: valgrind, sanitizers, and miri.
//!
//!   app <leak|uaf|uninit|overflow|race>
//!
//! Safe Rust's compile-time guarantees rule out four of the five seeded
//! defects outright: the borrow checker won't let a use-after-free or a
//! data race compile, and there's no way to name an uninitialised value or
//! index past a slice's bound without either a panic (bounds check) or an
//! explicit `unsafe` block. So `uaf`, `uninit`, `overflow`, and `race` each
//! print a fixed note and exit 64 -- the compile-time rejection *is* the
//! finding, reproduced here as a documented outcome rather than a crash.
//!
//! What safe Rust can still do is leak memory on purpose: `Box::leak`
//! consumes a `Box` and hands back a `'static` reference, deliberately
//! opting the allocation out of Rust's normal drop-based reclamation.
//! Nothing frees it -- valgrind's memcheck reports the same "definitely
//! lost" block it would for an equivalent C/C++ leak.

use std::env;
use std::hint::black_box;
use std::process::ExitCode;

const USAGE: &str = "usage: app <leak|uaf|uninit|overflow|race>";
const PREVENTED: &str = "rust: prevented at compile time — see chapter";

#[derive(Clone, Copy)]
enum Bug {
    Leak,
    Uaf,
    Uninit,
    Overflow,
    Race,
}

fn parse_bug(arg: &str) -> Result<Bug, String> {
    match arg {
        "leak" => Ok(Bug::Leak),
        "uaf" => Ok(Bug::Uaf),
        "uninit" => Ok(Bug::Uninit),
        "overflow" => Ok(Bug::Overflow),
        "race" => Ok(Bug::Race),
        other => Err(format!("unknown bug: {other}")),
    }
}

// leak: Box::leak is a safe API whose entire purpose is to opt an
// allocation out of automatic reclamation -- the program exits normally,
// but the 64 KiB Vec (and the Box that briefly held it) are never freed.
//
// black_box on the pointer is load-bearing, not decoration: without it,
// LLVM can see that nothing observable depends on the buffer's *contents*
// (only its statically-known length gets printed) and optimizes the
// entire allocation away, leak and all -- confirmed on this host by
// valgrind reporting zero bytes lost until the barrier was added.
fn leak() {
    let leaked: &'static mut Vec<u8> = Box::leak(Box::new(vec![0xAAu8; 65536]));
    leaked[0] = 0xAA; // touch it so the write isn't provably dead
    black_box(leaked.as_ptr());
    println!(
        "bugfarm: leak: leaked {} bytes via Box::leak, never reclaimed (intentional)",
        leaked.len()
    );
}

/// Returns `Ok(())` on the expressible bug (leak), or `Err(64)` for a bug
/// safe Rust prevents outright, or `Err(2)` for a usage error.
fn run(args: &[String]) -> Result<(), i32> {
    if args.len() != 1 {
        return Err(2);
    }
    let bug = parse_bug(&args[0]).map_err(|e| {
        eprintln!("app: {e}");
        2
    })?;
    match bug {
        Bug::Leak => {
            leak();
            Ok(())
        }
        Bug::Uaf | Bug::Uninit | Bug::Overflow | Bug::Race => {
            eprintln!("{PREVENTED}");
            Err(64)
        }
    }
}

fn main() -> ExitCode {
    let args: Vec<String> = env::args().skip(1).collect();
    match run(&args) {
        Ok(()) => ExitCode::SUCCESS,
        Err(2) => {
            eprintln!("{USAGE}");
            ExitCode::from(2)
        }
        Err(code) => ExitCode::from(code as u8),
    }
}
